src/python/pants/core/goals/test.py

```diff
@@ -204,6 +223,135 @@ class TestFieldSet(FieldSet, metaclass=ABCMeta):
+            This option will give the most fine-grained caching of test results, potentially
+            at the expense of performance on cache misses.
+        - Returning partitions containing multiple elements, to test multiple inputs per process.
+            This approach can make your tests more performant by sharing expensive setup/
+            teardown logic across many inputs, but results won't be cached as effectively.
+    """
+
+    @classmethod
+    def partition_per_input(cls, elements: Iterable[_FieldSetT]) -> Partitions[_FieldSetT]:
+        """Helper constructor for implementations that have a partition per input."""
+        return Partitions(
+            [(e.address.path_safe_spec, Partition((e,), e.address.spec)) for e in elements]
+        )
+
+
+@union
+class TestRequest:
+    """Base class for plugin types wanting to be run as part of `test`.
+
+    Plugins should define a new type which subclasses `TestRequest`, and set the appropriate class
+    variables. E.g.
+        class SuperTesterRequest(TestRequest):
+                name = SuperTesterSubsystem.options_scope
+                field_set_type = SuperTesterFieldSet
+
+    Then, define 4 `@rule`s:
+        1. A rule which takes an instance of your request type's `PartitionRequest` class property,
+            and returns a `Partitions` instance. E.g.
+                @rule
+                async def partition(
+                    request: SuperTesterRequest.PartitionRequest[SuperTesterFieldSet],
+                    subsystem: SuperTesterSubsystem,
+                ) -> Partitions[SuperTesterFieldSet]:
+                    if subsystem.skip:
+                        return Partitions()
+
+                    # One possible implementation
+                    return Partitions.partition_per_input(request.field_sets)
+
+        2. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestResult` instance. E.g.
+                @rule
+                async def super_test(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestResult:
+                    ...
+
+        3. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestDebugRequest` instance. E.g.
+                @rule
+                async def setup_super_test_debug(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestDebugRequest:
+                    ...
+
+        4. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestDebugAdapterRequest` instance. E.g.
+                @rule
+                async def setup_super_test_debug(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestDebugAdapterRequest:
```

Comment:
Since these are so similar its hard to grok their differences. I suggest making this a single element and talk about each of the required implementations.

Reply:
I opened #17070 for updating the docs in an `.md` file - mind if I leave off any polish for instructions until I tackle that one?

Reply:
sure thing